### ISSUE TITLE
Proper cleanup of thread-specific error data relies on particular sequence of key creation

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#5899]
I have a test binary that started complaining about memory leaks upon upgrading 1.1.0g to 1.1.0h:

```
Direct leak of 5256 byte(s) in 9 object(s) allocated from:
    #0 0x4f000c in __interceptor_malloc /x/eng/toolchain/prod/llvm_release_50/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:67:3
    #1 0x2b296ad3ae2e in CRYPTO_malloc (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libcrypto.so.1.1+0x1d5e2e)
    #2 0x2b296ad3ae9f in CRYPTO_zalloc (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libcrypto.so.1.1+0x1d5e9f)
    #3 0x2b296ad028f1 in ERR_get_state (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libcrypto.so.1.1+0x19d8f1)
    #4 0x2b296ad029b9 in ERR_clear_error (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libcrypto.so.1.1+0x19d9b9)
    #5 0x2b296af0ed10  (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libssl.so.1.1+0x5dd10)
    #6 0x2b296af0f386 in ossl_statem_accept (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libssl.so.1.1+0x5e386)
    #7 0x2b296aefd50e in SSL_do_handshake (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libssl.so.1.1+0x4c50e)
    #8 0x2b296aefd369 in SSL_accept (/x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/bedrock/internal/linux64.debug.asan/utest-l/lib/libssl.so.1.1+0x4c369)
    #9 0x55285a in ListenerTestSuite::serverSslHandshake(int) /x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/./src/tests/syslog_fwd.ut:569:23
    #10 0x552163 in ListenerTestSuite::listenerThread(void*) /x/eng/bbrtp/users/jwalch/update_11_4820913_1804031335/apps/lib/libmsg_fwd/./src/tests/syslog_fwd.ut:591:35
    #11 0x2b296b5b6e0f in start_thread /u/hiralo/TGB/git/depot.prod.rats.tools.ustools_llvm/epi/build/x86_64-redhat-rhel6/glibc-2.12.1/nptl/pthread_create.c:301

SUMMARY: AddressSanitizer: 5256 byte(s) leaked in 9 allocation(s).
```

I ran under GDB and encountered a snippet that had me scratching my head for some time:

```
(gdb) b pthread_setspecific if $rdi == 1
Breakpoint 1 at 0x434ce0
(gdb) b err_delete_thread_state
Function "err_delete_thread_state" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (err_delete_thread_state) pending.
(gdb) run
[...]
[New LWP 5975]
[Switching to LWP 5975]

Breakpoint 1, __pthread_setspecific (key=1, value=0x7fff98000a20) at pthread_setspecific.c:29
29      in pthread_setspecific.c
(gdb) bt
#0  __pthread_setspecific (key=1, value=0x7fff98000a20) at pthread_setspecific.c:29
#1  0x00007ffff78960bc in CRYPTO_THREAD_set_local (key=0x7ffff798d654 <err_thread_local>, val=0x7fff98000a20) at crypto/threads_pthread.c:127
[...]
#10 0x00007ffff6fd5e10 in start_thread (arg=0x7fffa225c710) at pthread_create.c:301
#11 0x00007ffff6376b7d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:115
(gdb) fin
Run till exit from #0  __pthread_setspecific (key=1, value=0x7fff98000a20) at pthread_setspecific.c:29
0x00007ffff78960bc in CRYPTO_THREAD_set_local (key=0x7ffff798d654 <err_thread_local>, val=0x7fff98000a20) at crypto/threads_pthread.c:127
127         if (pthread_setspecific(*key, val) != 0)
Value returned is $1 = 0
(gdb) c
Continuing.

Breakpoint 2, err_delete_thread_state () at crypto/err/err.c:639
639         ERR_STATE *state = CRYPTO_THREAD_get_local(&err_thread_local);
(gdb) bt
#0  err_delete_thread_state () at crypto/err/err.c:639
[...]
#3  0x00007ffff6fd5bd2 in __nptl_deallocate_tsd () at pthread_create.c:154
#4  0x00007ffff6fd5e1e in start_thread (arg=0x7fffa225c710) at pthread_create.c:308
#5  0x00007ffff6376b7d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:115
(gdb) s
CRYPTO_THREAD_get_local (key=0x7ffff798d654 <err_thread_local>) at crypto/threads_pthread.c:122
122         return pthread_getspecific(*key);
(gdb) p *key
$2 = 1
(gdb) fin
Run till exit from #0  CRYPTO_THREAD_get_local (key=0x7ffff798d654 <err_thread_local>) at crypto/threads_pthread.c:122
0x00007ffff77e4a61 in err_delete_thread_state () at crypto/err/err.c:639
639         ERR_STATE *state = CRYPTO_THREAD_get_local(&err_thread_local);
Value returned is $3 = (void *) 0x0
```
The set succeeded, but the get returns NULL? Didn't make any sense.

Then I realized that err_delete_thread_state was being called as a consequence of the handler of the key=2 data (ossl_init_thread_stop). Since it's in order, the key=1 thread-specific data was long gone.

Basically, in order for the data associated with err_thread_local to be properly cleaned up, that key has to be registered (via CRYPTO_THREAD_init_local & then pthread_key_create) AFTER threadstopkey (which registers the handler). I believe that was always the case prior to https://github.com/openssl/openssl/commit/aef84bb4efbddfd95d042f3f5f1d362ed7d4faeb, 
but is not necessarily the case afterward.

To avoid this sort of issue, the registration of err_thread_local should have its own accompanying cleanup handler. I modified the code to do this (basically just calling ERR_STATE_free on the pointer passed to the callback), and it resolved the leak (at least according to ASAN, in this application).

**Comment by bernd-edlinger** [comment]
What happens if you call `OPENSSL_init_crypto(0, NULL);`
in the main thread?

**Comment by bernd-edlinger** [comment]
I mean before any other thread is stared.

**Comment by jwalch** [comment]
I'm not too concerned with what application code changes my resolve the issue. Ideally an application should be able to link against a new security release without any code changes and not see a new perceived memory leak.

The other side of it is that a selling point of 1.1.x over 1.0.2 is that explicit initialization / de-initialization of the library is no longer necessary. It's even mentioned in the man page:

> As of version 1.1.0 OpenSSL will automatically allocate all resources that it needs so no explicit initialisation is required. Similarly it will also automatically deinitialise as required.

If the documentation didn't make such promises about the library handling its own resource management, having expectations on the client would more reasonable. (Say, for example, this type of issue had been introduced in a 1.0.2 security release.)

I'm surprised you see the issue in 1.1.0g. My test application does NOT report the leak against that version.

I should note that I don't believe https://github.com/openssl/openssl/commit/aef84bb4efbddfd95d042f3f5f1d362ed7d4faeb is fundamentally buggy. I understand the leak it addresses (and to which you refer). This seems to me to be one of those pernicious issues that arise when undocumented dependencies exist in code; "good" changes *expose* existing bugs by modifying a behavior that was never guaranteed (but which the "bad" code relied upon for correct operation).

IMO the real bug here is that the memory in question is associated with a different "key" than the one it is cleaned up with.

**Comment by bernd-edlinger** [comment]
I think this was always a latent bug, apparently something changed between 1.1.0g and 1.1.0h
that made your thread invoke ERR_get_error before OPENSSL_init_crypto.
While it previously was the other way round.
I am a bit concerned about this comment in crypto/init.c:
```
DEFINE_RUN_ONCE_STATIC(ossl_init_base)
{
#ifdef OPENSSL_INIT_DEBUG
    fprintf(stderr, "OPENSSL_INIT: ossl_init_base: Setting up stop handlers\n");
#endif
    /*
     * We use a dummy thread local key here. We use the destructor to detect
     * when the thread is going to stop (where that feature is available)
     */
    CRYPTO_THREAD_init_local(&threadstopkey, ossl_init_thread_stop_wrap);
#ifndef OPENSSL_SYS_UEFI
    atexit(OPENSSL_cleanup);
#endif
    if ((init_lock = CRYPTO_THREAD_lock_new()) == NULL)
        return 0;
    OPENSSL_cpuid_setup();

    /*
     * BIG FAT WARNING!
     * Everything needed to be ini
DEFINE_RUN_ONCE_STATIC(ossl_init_base)
{
#ifdef OPENSSL_INIT_DEBUG
    fprintf(stderr, "OPENSSL_INIT: ossl_init_base: Setting up stop handlers\n");
#endif
    /*
     * We use a dummy thread local key here. We use the destructor to detect
     * when the thread is going to stop (where that feature is available)
     */
    CRYPTO_THREAD_init_local(&threadstopkey, ossl_init_thread_stop_wrap);
#ifndef OPENSSL_SYS_UEFI
    atexit(OPENSSL_cleanup);
#endif
    if ((init_lock = CRYPTO_THREAD_lock_new()) == NULL)
        return 0;
    OPENSSL_cpuid_setup();

    /*
     * BIG FAT WARNING!
     * Everything needed to be initialized in this function before threads
     * come along MUST happen before base_inited is set to 1, or we will
     * see race conditions.
     */
    base_inited = 1;
tialized in this function before threads
     * come along MUST happen before base_inited is set to 1, or we will
     * see race conditions.
     */
    base_inited = 1;
```

Because what this bug exposes is exactly a race between two threads initializing
OPENSSL_init_crypto.  Thread sanitizer reports a race here.
